src/Bicep.LangServer.IntegrationTests/TelemetryTests.cs

```diff
@@ -0,0 +1,142 @@
+            IDictionary<string, string> properties = new Dictionary<string, string>
+            {
+                { "name", prefix },
+                { "type", "Microsoft.ContainerService/managedClusters@2021-03-01" }
+            };
+
+            TaskCompletionSource<TelemetryEventParams> telemetryReceived = await ResolveCompletionAsync(text, prefix, TelemetryConstants.EventNames.ResourceBodySnippetInsertion, properties);
+            TelemetryEventParams telemetryEventParams = await IntegrationTestHelper.WithTimeoutAsync(telemetryReceived.Task);
+            IDictionary<string, object> extensionData = telemetryEventParams.ExtensionData;
+            ICollection<string> keys = extensionData.Keys;
+
+            keys.Count.Should().Be(2);
+            keys.Should().Contain("eventName");
+            keys.Should().Contain("properties");
+            extensionData["eventName"].ToString().Should().Be(TelemetryConstants.EventNames.ResourceBodySnippetInsertion);
+            extensionData["properties"].ToString().Should().BeEquivalentToIgnoringNewlines(expectedProperties);
+        }
+
+        [TestMethod]
+        public async Task VerifyModuleBodySnippetInsertionFiresTelemetryEvent()
+        {
+            string text = "module foo 'test.bicep' = ";
+            IDictionary<string, string> properties = new Dictionary<string, string>
+            {
+                { "name", "{}" }
+            };
+
+            TaskCompletionSource<TelemetryEventParams> telemetryReceived = await ResolveCompletionAsync(text, "{}", TelemetryConstants.EventNames.ModuleBodySnippetInsertion, properties);
+            TelemetryEventParams telemetryEventParams = await IntegrationTestHelper.WithTimeoutAsync(telemetryReceived.Task);
+            IDictionary<string, object> extensionData = telemetryEventParams.ExtensionData;
+            ICollection<string> keys = extensionData.Keys;
+
+            keys.Count.Should().Be(2);
+            keys.Should().Contain("eventName");
+            keys.Should().Contain("properties");
+            extensionData["eventName"].ToString().Should().Be(TelemetryConstants.EventNames.ModuleBodySnippetInsertion);
+            extensionData["properties"].ToString().Should().BeEquivalentToIgnoringNewlines(@"{
+  ""name"": ""{}""
+}");
+        }
+
+        private async Task<TaskCompletionSource<TelemetryEventParams>> ResolveCompletionAsync(string text, string prefix, string eventName, IDictionary<string, string> properties)
+        {
+            var fileSystemDict = new Dictionary<Uri, string>();
+            var telemetryReceived = new TaskCompletionSource<TelemetryEventParams>();
+
+            var client = await IntegrationTestHelper.StartServerWithClientConnectionAsync(
+                TestContext,
+                options =>
+                {
+                    options.OnTelemetryEvent(telemetry => telemetryReceived.SetResult(telemetry));
```

Comment:
I think you can use the typed handler here to avoid the dictionary serialization that `TelemetryEventParams` does:

```csharp
options.OnTelemetryEvent<BicepTelemetryEvent>(telemetry => telemetryReceived.SetResult(telemetry));
```

Reply:
Updated here:
https://github.com/Azure/bicep/commit/62d6763f06b3d53bb8189b517dab0567a37fd0bd